File: Motocitizen/src/motocitizen/Activity/DetailVolunteersFragment.java
package motocitizen.Activity;

import android.app.Activity;
import android.app.DialogFragment;
import android.app.Fragment;
import android.app.FragmentTransaction;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.Toast;

import java.util.HashMap;
import java.util.Map;

import motocitizen.app.general.Accident;
import motocitizen.app.general.AccidentVolunteer;
import motocitizen.app.general.AccidentsGeneral;
import motocitizen.main.R;
import motocitizen.network.JsonRequest;
import motocitizen.network.OnwayRequest;
import motocitizen.startup.Startup;

import static motocitizen.app.general.AccidentsGeneral.getDelimiterRow;

public class DetailVolunteersFragment extends AccidentDetailsFragments {

    public static final int DIALOG_ONWAY_CONFIRM = 1;
    public static final int DIALOG_ACC_NOT_ACTUAL = 2;

    private ImageButton onwayButton;
    private View toMap;
    private View onwayContent;
    private View inplaceContent;

    // TODO: Rename and change types and number of parameters
    public static DetailVolunteersFragment newInstance(int accID, String userName) {
        DetailVolunteersFragment fragment = new DetailVolunteersFragment();
        Bundle args = new Bundle();
        args.putInt(ACCIDENT_ID, accID);
        args.putString(USER_NAME, userName);
        fragment.setArguments(args);
        return fragment;
    }

    public DetailVolunteersFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View viewMain = inflater.inflate(R.layout.fragment_detail_volunteers, container, false);

        onwayButton = (ImageButton) viewMain.findViewById(R.id.onway_button);
        onwayButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showDialog(DIALOG_ONWAY_CONFIRM);
            }
        });

        onwayContent = viewMain.findViewById(R.id.acc_onway_table);
        inplaceContent = viewMain.findViewById(R.id.acc_inplace_table);
        toMap = viewMain.findViewById(R.id.details_to_map_button);
        toMap.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ((AccidentDetailsActivity) getActivity()).jumpToMap();
            }
        });
        update();
        return viewMain;
    }

    protected void update() {
        Accident accident = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();

        if (accident != null) {
            setupAccess();

            ViewGroup vg_onway = (ViewGroup) onwayContent;
            ViewGroup vg_inplace = (ViewGroup) inplaceContent;
            vg_onway.setVisibility(View.INVISIBLE);
            vg_inplace.setVisibility(View.INVISIBLE);
            vg_onway.removeAllViews();
            vg_inplace.removeAllViews();
            for (int i : accident.getSortedVolunteersKeys()) {
                AccidentVolunteer current = accident.volunteers.get(i);
                switch (current.getStatus()) {
                    case ONWAY:
                        if (vg_onway.getVisibility() == View.INVISIBLE) {
                            vg_onway.setVisibility(View.VISIBLE);
                            vg_onway.addView(getDelimiterRow(getActivity(), "В пути"));
                        }
                        vg_onway.addView(current.createRow(getActivity()));
                        break;
                    case INPLACE:
                        if (vg_onway.getVisibility() == View.INVISIBLE) {
                            vg_onway.setVisibility(View.VISIBLE);
                            vg_onway.addView(getDelimiterRow(getActivity(), "На месте"));
                        }
                        vg_onway.addView(current.createRow(getActivity()));
                        break;
                    case LEAVE:
                        if (vg_onway.getVisibility() == View.INVISIBLE) {
                            vg_onway.setVisibility(View.VISIBLE);
                            vg_onway.addView(getDelimiterRow(getActivity(), "Были"));
                        }
                        vg_onway.addView(current.createRow(getActivity()));
                        break;
                }
            }
        } else {
            showDialog(DIALOG_ACC_NOT_ACTUAL);
        }
    }

    public void notifyDataSetChanged() {
        update();
//  ListAdapter
//        adapter.notifyDataSetChanged();
    }

    private void setupAccess() {
        Accident accident = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();
        if (accident.getId() == prefs.getOnWay() || accident.getId() == AccidentsGeneral.getInplaceID() || !AccidentsGeneral.auth.isAuthorized() || !accident.isActive()) {
            onwayButton.setVisibility(View.INVISIBLE);
        } else {
            onwayButton.setVisibility(View.VISIBLE);
        }
    }

    void showDialog(int type) {
        mStackLevel++;
        FragmentTransaction ft = getActivity().getFragmentManager().beginTransaction();
        Fragment prev = getActivity().getFragmentManager().findFragmentByTag("dialog");
        if (prev != null) {
            ft.remove(prev);
        }
        ft.addToBackStack(null);
        ft.commit();
        Activity act = getActivity();

        switch (type) {
            case DIALOG_ONWAY_CONFIRM:
                DialogFragment onwayConfirm = ConfirmDialog.newInstance(
                        act.getString(R.string.title_dialog_onway_confirm),
                        act.getString(android.R.string.yes),
                        act.getString(android.R.string.no));
                onwayConfirm.setTargetFragment(this, DIALOG_ONWAY_CONFIRM);
                onwayConfirm.show(getFragmentManager().beginTransaction(), "dialog");
                break;
            case DIALOG_ACC_NOT_ACTUAL:
                DialogFragment dialogFrag = ConfirmDialog.newInstance(
                        act.getString(R.string.title_dialod_acc_not_actual),
                        act.getString(android.R.string.ok),
                        "");
                dialogFrag.setTargetFragment(this, DIALOG_ACC_NOT_ACTUAL);
                dialogFrag.show(getFragmentManager().beginTransaction(), "dialog");
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case DIALOG_ONWAY_CONFIRM:
                if (resultCode == Activity.RESULT_OK) {
                    sendOnway();
                } else if (resultCode == Activity.RESULT_CANCELED) {
                    // After Cancel code.
                }
                break;
            case DIALOG_ACC_NOT_ACTUAL:
                getActivity().finish();
                break;
        }
    }

    private void sendOnway() {
        if (Startup.isOnline()) {
            int currentId = AccidentsGeneral.getCurrentPointID();
            AccidentsGeneral.setOnWay(currentId);
            Map<String, String> post = new HashMap<>();
            post.put("login", AccidentsGeneral.auth.getLogin());
            post.put("passhash", AccidentsGeneral.auth.makePassHash());
            post.put("id", String.valueOf(currentId));
            JsonRequest request = new JsonRequest("mcaccidents", "onway", post, "", true);
            if (request != null) {
                (new OnwayRequest((AccidentDetailsActivity) getActivity(), currentId)).execute(request);
            }
        } else {
            Toast.makeText(getActivity(), getActivity().getString(R.string.inet_not_available), Toast.LENGTH_LONG).show();
        }
    }
}

File: Motocitizen/src/motocitizen/ui/fragments/DetailVolunteersFragment.kt
package motocitizen.ui.fragments

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import motocitizen.content.Content
import motocitizen.content.accident.Accident
import motocitizen.datasources.network.requests.CancelOnWayRequest
import motocitizen.datasources.network.requests.OnWayRequest
import motocitizen.datasources.preferences.Preferences
import motocitizen.main.R
import motocitizen.ui.activity.AccidentDetailsActivity
import motocitizen.ui.activity.AccidentDetailsActivity.Companion.ACCIDENT_ID_KEY
import motocitizen.ui.dialogs.details.ConfirmDialog
import motocitizen.ui.rows.volunteer.VolunteerRow
import motocitizen.utils.hide
import motocitizen.utils.show
import org.jetbrains.anko.runOnUiThread

class DetailVolunteersFragment : FragmentForAccident() {
    companion object {
        private const val DIALOG_ON_WAY_CONFIRM = 1
        private const val DIALOG_CANCEL_ON_WAY_CONFIRM = 2
    }

    private lateinit var rootView: View
    private lateinit var confirmButton: ImageButton
    private lateinit var cancelButton: ImageButton
    private lateinit var disabledButton: ImageButton
    private lateinit var content: ViewGroup
    private lateinit var toMapButton: View

    private lateinit var accident: Accident

    override fun setAccident(accident: Accident) {
        this.accident = accident
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_detail_volunteers, container, false)
        bindViews()
        disabledButton.isEnabled = false
        setUpListeners()
        update()
        return rootView
    }

    private fun bindViews() {
        confirmButton = rootView.findViewById(R.id.onway_button)
        cancelButton = rootView.findViewById(R.id.onway_cancel_button)
        disabledButton = rootView.findViewById(R.id.onway_disabled_button)
        content = rootView.findViewById(R.id.acc_onway_table)
        toMapButton = rootView.findViewById(R.id.details_to_map_button)
    }

    private fun setUpListeners() {
        confirmButton.setOnClickListener { showOnWayDialog() }
        cancelButton.setOnClickListener { showCancelDialog() }
        toMapButton.setOnClickListener { (activity as AccidentDetailsActivity).toMap() }
    }

    private fun update() {
        setupAccess()
        content.removeAllViews()
        accident.volunteers.forEach {
            try {
                content.addView(VolunteerRow(activity, it))
            } catch (e: IllegalArgumentException) {
                //todo fuckup
            }
        }
    }

    private fun setupAccess() = runOnUiThread {
        cancelButton.apply { if (canShowCancel()) show() else hide() }
        confirmButton.apply { if (canShowConfirm()) show() else hide() }
        disabledButton.apply { if (canShowDisabled()) show() else hide() }
    }

    private fun canShowCancel() = accident.isActive() && accident != Content.inPlace && accident.id == Preferences.onWay

    private fun canShowConfirm() = accident.isActive() && accident != Content.inPlace && accident.id != Preferences.onWay

    private fun canShowDisabled() = accident.isActive() && accident == Content.inPlace

    private fun showOnWayDialog() {
        val dialog = ConfirmDialog()
        dialog.title = activity.getString(R.string.title_dialog_onway_confirm)
        dialog.setTargetFragment(this, DIALOG_ON_WAY_CONFIRM)
        dialog.show(fragmentManager, "dialog")

    }

    private fun showCancelDialog() {
        val dialog = ConfirmDialog()
        dialog.title = activity.getString(R.string.title_dialog_cancel_onway_confirm)
        dialog.setTargetFragment(this, DIALOG_CANCEL_ON_WAY_CONFIRM)
        dialog.show(fragmentManager, "dialog")
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (resultCode == Activity.RESULT_CANCELED) return
        when (requestCode) {
            DIALOG_ON_WAY_CONFIRM        -> sendOnWay()
            DIALOG_CANCEL_ON_WAY_CONFIRM -> sendCancelOnWay()
        }
    }

    private fun sendOnWay() {
        Preferences.onWay = accident.id
        OnWayRequest(accident.id) { setupAccess() }.call() //todo
    }

    private fun sendCancelOnWay() {
        Preferences.onWay = 0
        CancelOnWayRequest(accident.id) { setupAccess() }.call() //todo
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(ACCIDENT_ID_KEY, accident.id)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) return
        accident = Content[savedInstanceState.getInt(ACCIDENT_ID_KEY)]!!
    }
}
